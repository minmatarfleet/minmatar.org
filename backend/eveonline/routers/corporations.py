from typing import List

from ninja import ModelSchema, Router, Schema

from authentication import AuthBearer, requires_permission
from eveonline.models import EveCorporation, EveCorporationApplication

router = Router(tags=["Corporations"])


class CorporationResponse(ModelSchema):
    class Meta:
        model = EveCorporation
        fields = "__all__"


class CorporationApplicationResponse(Schema):
    status: str
    description: str
    corporation_id: int


class CreateCorporationRequest(Schema):
    corporation_id: int


@router.get(
    "/corporations",
    response=List[CorporationResponse],
    auth=AuthBearer(),
    summary="Get all corporations",
)
def get_corporations(request):
    return EveCorporation.objects.all()


@router.get(
    "/corporations/{corporation_id}",
    response=CorporationResponse,
    auth=AuthBearer(),
    summary="Get a corporation by ID",
)
def get_corporation_by_id(request, corporation_id: int):
    return EveCorporation.objects.get(corporation_id=corporation_id)


@router.post(
    "/corporations",
    response=CorporationResponse,
    auth=AuthBearer(),
    summary="Create a corporation",
)
@requires_permission("eveonline.add_evecorporation")
def create_corporation(request, payload: CreateCorporationRequest):
    return EveCorporation.objects.create(**payload.dict())


@router.get(
    "/corporations/applications",
    response=List[CorporationApplicationResponse],
    auth=AuthBearer(),
    summary="Get corporation applications",
)
@requires_permission("eveonline.view_evecorporationapplication")
def get_corporation_applications(request):
    return EveCorporationApplication.objects.all()
