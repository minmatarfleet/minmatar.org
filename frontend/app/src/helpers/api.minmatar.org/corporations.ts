import type { Corporation, CorporationType, CorporationInfo } from '@dtypes/api.minmatar.org'
import { get_error_message } from '@helpers/string'

const API_ENDPOINT =  `${import.meta.env.API_URL}/api/eveonline/corporations`

export async function get_all_corporations(corporation_type:CorporationType) {
    const headers = {
        'Content-Type': 'application/json'
    }

    const ENDPOINT = `${API_ENDPOINT}/corporations?corporation_type=${corporation_type}`

    console.log(`Requesting: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `GET ${ENDPOINT}`
            ));
        }

        return await response.json() as Corporation[];
    } catch (error) {
        throw new Error(`Error fetching corporations: ${error.message}`);
    }
}

export async function get_corporation_by_id(access_token:string, id:number) {
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${access_token}`
    }

    const ENDPOINT = `${API_ENDPOINT}/corporations/${id}`

    console.log(`Requesting: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `GET ${ENDPOINT}`
            ));
        }

        return await response.json() as Corporation;
    } catch (error) {
        throw new Error(`Error fetching corporation: ${error.message}`);
    }
}

export async function get_corporation_info(id:number) {
    const headers = {
        'Content-Type': 'application/json',
    }

    const ENDPOINT = `${API_ENDPOINT}/corporations/${id}/info`

    console.log(`Requesting: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `GET ${ENDPOINT}`
            ));
        }

        return await response.json() as CorporationInfo;
    } catch (error) {
        throw new Error(`Error fetching corporation: ${error.message}`);
    }
}