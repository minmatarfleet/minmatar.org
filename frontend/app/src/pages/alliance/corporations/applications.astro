---
import { i18n } from '@helpers/i18n'
const { lang, t, translatePath } = i18n(Astro.url)

import { HTTP_403_Forbidden } from '@helpers/http_responses'

import type { User } from '@dtypes/jwt'
import * as jose from 'jose'

const auth_token = Astro.cookies.has('auth_token') ? (Astro.cookies.get('auth_token')?.value as string) : null
const user = auth_token ? jose.decodeJwt(auth_token) as User : false
const is_superuser = (user ? user?.is_superuser : false)

import { get_user_permissions } from '@helpers/permissions'
const user_permissions = (user ? await get_user_permissions(user?.username) : [])

const can_view_page = is_superuser || user_permissions.includes('applications.change_evecorporationapplication')

if (!can_view_page)
    return HTTP_403_Forbidden()

import { get_applications_data } from '@helpers/pages/applications'
import type { ApplicationsData } from '@helpers/pages/applications'

let applications_data:ApplicationsData = {}
let data_fetching_error

try {
    applications_data = await get_applications_data(auth_token as string, lang)
} catch (error) {
    data_fetching_error = error.message
}

const {
    corporations_applications = [],
    corporations_unfiltered = [],
    corporations_applications_count = [],
    total_applications = 0,
    corporations_options = [],
} = applications_data

const CORPORATIONS_APPLICATIONS_PARTIAL_URL = translatePath('/partials/corporations_applications_component')

import Viewport from '@layouts/Viewport.astro';

import PageDefault from '@components/page/PageDefault.astro';
import PageTitle from '@components/page/PageTitle.astro';

import FlexInline from '@components/compositions/FlexInline.astro';
import Flexblock from '@components/compositions/Flexblock.astro';

import CorporationApplicationsList from '@components/blocks/CorporationApplicationsList.astro';
import TextBox from '@components/layout/TextBox.astro';
import Input from '@components/blocks/Input.astro';
import Select from '@components/blocks/Select.astro';
import ErrorRefetch from '@components/blocks/ErrorRefetch.astro';
import Button from '@components/blocks/Button.astro';

import MagnifierIcon from '@components/icons/buttons/MagnifierIcon.astro';

const page_title = t('corporations.applications.page_title')
---

<Viewport
    title={page_title}
    components={{
        alert_dialog: true,
        confirm_dialog: true,
    }}
>
    <PageDefault
        cover={{
            image: "/images/alliance-cover.jpg",
            image_990: "/images/alliance-cover.jpg",
            animated: false,
            scrollable: true,
            overlay: true
        }}
        x-data={`{
            global_collapse: false,
            corporations_applications: ${JSON.stringify(corporations_applications)},
            corporations_unfiltered: ${JSON.stringify(corporations_unfiltered)},
            corporations_filtered: ${JSON.stringify(corporations_unfiltered)},
            corporations_filtered_applications_count: ${JSON.stringify(corporations_applications_count)},
            total_applications_filtered: ${total_applications},
            total_applications_unfiltered: ${total_applications},
            corporations_options: ${JSON.stringify(corporations_options)},
            character_name: '',
            corporation_filter: '-1',
            filter() {
                if (this.character_name === '' && this.corporation_filter === '-1') {
                    this.corporations_filtered = this.corporations_unfiltered
                    this.corporations_filtered_applicationss_count = ${JSON.stringify(corporations_applications_count)}
                    this.total_applications_filtered = this.total_applications_unfiltered
                    return
                }

                this.corporations_filtered = {}
                this.corporations_filtered_applicationss_count = {}
                this.total_applications_filtered = 0

                this.corporations_applications.forEach( (corporation) => {
                    if (this.corporation_filter !== '-1' && parseInt(this.corporation_filter) !== corporation.corporation_id) {
                        this.corporations_filtered[corporation.corporation_id] = []
                        this.corporations_filtered_applications_count[corporation.corporation_id] = 0
                        return true
                    }

                    const applications = corporation.applications.filter( (application) => application.character_name.toLowerCase().includes(this.character_name.toLowerCase()) )
                    
                    if (applications.length > 0) {
                        this.total_applications_filtered += applications.length
                        this.corporations_filtered[corporation.corporation_id] = applications.map( (application) => application.id )
                        this.corporations_filtered_applications_count[corporation.corporation_id] = applications.length
                    }
                })
            },
        }`}
    >
        <FlexInline slot="header" justification='space-between' class="[ w-full ]">
            <PageTitle>
                {page_title}
            </PageTitle>
            <FlexInline>
                <Input type="text"
                    placeholder={t('filter')}
                    x-model="character_name"
                    x-on:keyup="filter()"
                >
                    <MagnifierIcon slot="icon" />
                </Input>
                
                <div class="[ hidden 2xl:block ]">
                    <Select
                        x-model="corporation_filter"
                        width='280px'
                        x-on:change="filter()"
                    >
                        <option selected value="-1">{t('corporations')}</option>
                        <template x-for="corporation in corporations_options">
                            <option x-bind:value="corporation.value" x-text="corporation.label" />
                        </template>
                    </Select>
                </div>
            </FlexInline>
        </FlexInline>

        <FlexInline slot="subheader">
            <Button
                href={translatePath('/alliance/corporations/applications/records')}
                size='sm'
            >
                {t('view_records')}
            </Button>
        </FlexInline>

        <Flexblock gap="var(--space-xl)">
            <TextBox>
                <p>{t('corporations.applications.leading_text')}</p>
            </TextBox>

            {data_fetching_error ?
                <ErrorRefetch
                    args={{
                        partial: CORPORATIONS_APPLICATIONS_PARTIAL_URL,
                        message: data_fetching_error,
                        delay: 0,
                    }}
                />
                :
                <CorporationApplicationsList corporations={corporations_applications} />
            }
        </Flexblock>
    </PageDefault>
</Viewport>