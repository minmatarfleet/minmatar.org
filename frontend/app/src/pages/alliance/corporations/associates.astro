---
import { i18n } from '@helpers/i18n'
const { lang, t, translatePath } = i18n(Astro.url)

import type { User } from '@dtypes/jwt'
import * as jose from 'jose'

const auth_token = Astro.cookies.has('auth_token') ? (Astro.cookies.get('auth_token')?.value as string) : null
const user = auth_token ? jose.decodeJwt(auth_token) as User : null
const user_id = user?.user_id ?? null

import { get_associates_data } from '@helpers/pages/associates'
import type { AssociatesData } from '@helpers/pages/associates'

let associates_data:AssociatesData = {}
let data_fetching_error:string | null = null

try {
    associates_data = await get_associates_data(auth_token, lang, user_id)
} catch (error) {
    data_fetching_error = error
}

const {
    corporations = [],
    user_corporation_id = null,
} = associates_data

const CORPORATIONS_LIST_PARTIAL_URL = `${translatePath('/partials/corporations_list_component')}?corporation_type=associate`

import Viewport from '@layouts/Viewport.astro';

import PageDefault from '@components/page/PageDefault.astro';
import PageTitle from '@components/page/PageTitle.astro';

import Flexblock from '@components/compositions/Flexblock.astro';
import FlexInline from '@components/compositions/FlexInline.astro';

import CorporationsList from '@components/blocks/CorporationsList.astro';
import ErrorRefetch from '@components/blocks/ErrorRefetch.astro';

const page_title = t('corporations.associates.page_title');
const page_description = t('corporations.associates.leading_text');
---

<Viewport
    title={page_title}
    meta_description={page_description}
    components={{
        alert_dialog: true,
        confirm_dialog: true,
    }}
>
    <PageDefault
        wide={true}
        cover={{
            image: "/images/alliance-cover.jpg",
            image_990: "/images/alliance-cover-990.jpg",
            alt: t('alliance.cover_alt'),
            animated: false,
            scrollable: true,
            overlay: true
        }}
        x-data={`{
            corporation_name: '',
            corporations_count: ${corporations.length}
        }`}
    >
        <FlexInline slot="header" justification='space-between' class="[ w-full ]">
            <FlexInline>
                <Flexblock gap="var(--space-3xs)">
                    <PageTitle>
                        {page_title}
                    </PageTitle>
                    <small x-text={`\`\${corporations_count} \${corporations_count != 1 ? '${t('corporations').toLowerCase()}' : '${t('corporation').toLowerCase()}'}\``}>
                        {corporations.length} {corporations.length != 1 ? t('corporations').toLowerCase() : t('corporation').toLowerCase()}
                    </small>
                </Flexblock>
            </FlexInline>
        </FlexInline>

        <Flexblock gap="var(--space-xl)">
            {data_fetching_error ?
                <ErrorRefetch
                    args={{
                        partial: CORPORATIONS_LIST_PARTIAL_URL,
                        message: data_fetching_error,
                        delay: 0,
                    }}
                />
                :
                <CorporationsList
                    corporations={corporations}
                    user_corporation_id={user_corporation_id ?? 0}
                    disable_applications={true}
                />
            }
        </Flexblock>
	</PageDefault>
</Viewport>

<style lang="scss">
    h3 {
        margin-block-start: var(--space-l);
    }
</style>