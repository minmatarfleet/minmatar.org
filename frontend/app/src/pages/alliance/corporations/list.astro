---
import { i18n } from '@helpers/i18n'
const { lang, t, translatePath } = i18n(Astro.url)

import type { User } from '@dtypes/jwt'
import * as jose from 'jose'

const auth_token = Astro.cookies.has('auth_token') ? (Astro.cookies.get('auth_token')?.value as string) : null
const user = auth_token ? jose.decodeJwt(auth_token) as User : null
const user_id = user?.user_id ?? null

import { get_corporations_list_data } from '@helpers/pages/corporations_list'
import type { CorporationsListData } from '@helpers/pages/corporations_list'

let corporations_list_data:CorporationsListData = {}
let data_fetching_error:string | null = null

try {
    corporations_list_data = await get_corporations_list_data(auth_token, lang, user_id)
} catch (error) {
    data_fetching_error = error
}

const {
    corporations = [],
    user_corporation_id = null,
} = corporations_list_data

const CORPORATIONS_SLIDER_PARTIAL_URL = translatePath('/partials/corporations_slider_component/')

import Viewport from '@layouts/Viewport.astro';

import PageLanding from '@components/page/PageLanding.astro';

import CorporationsSlider from '@components/blocks/CorporationsSlider.astro';
import ErrorRefetch from '@components/blocks/ErrorRefetch.astro';
import VerticalCenter from '@components/blocks/VerticalCenter.astro';

const page_title = t('corporations.list.page_title');
---

<Viewport
    title={page_title}
    meta_description={t('corporations.list.meta_description')}
    components={{
        alert_dialog: true,
        confirm_dialog: true,
    }}
>
    <PageLanding
        wide={true}
        fullscreen={true}
        cover={{
            image: "/images/corporations-cover.jpg",
            image_990: "/images/corporations-cover-990.jpg",
            alt: t('corporations.cover_alt'),
            animated: false,
            scrollable: true,
            overlay: true
        }}
        x-data={`{
            corporation_name: '',
        }`}
    >
        <VerticalCenter class="[ min-h-[90vh] ]" gap="var(--space-xl)">
            {data_fetching_error ?
                <ErrorRefetch
                    args={{
                        partial: CORPORATIONS_SLIDER_PARTIAL_URL,
                        message: data_fetching_error,
                        delay: 0,
                    }}
                />
                :
                <CorporationsSlider
                    corporations={corporations}
                    user_corporation_id={user_corporation_id ?? 0}
                />
            }
        </VerticalCenter>
	</PageLanding>
</Viewport>

<style lang="scss">
    h3 {
        margin-block-start: var(--space-l);
    }
</style>