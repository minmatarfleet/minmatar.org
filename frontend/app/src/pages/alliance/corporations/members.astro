---
import { i18n } from '@helpers/i18n'
const { lang, t, translatePath } = i18n(Astro.url)

import { HTTP_403_Forbidden } from '@helpers/http_responses'

import type { User } from '@dtypes/jwt'
import * as jose from 'jose'

const auth_token = Astro.cookies.has('auth_token') ? (Astro.cookies.get('auth_token')?.value as string) : null
const user = auth_token ? jose.decodeJwt(auth_token) as User : false
const is_superuser = (user ? user?.is_superuser : false)

import { get_user_permissions } from '@helpers/permissions'
const user_permissions = (user ? await get_user_permissions(user?.username) : [])

const can_view_page = is_superuser || user_permissions.includes('eveonline.view_evecorporation')

if (!can_view_page)
    return HTTP_403_Forbidden()

import { get_corporations_members_data } from '@helpers/pages/corporations_members'
import type { CorporationsMembersData } from '@helpers/pages/corporations_members'

let corporations_members_data:CorporationsMembersData = {}
let data_fetching_error

try {
    corporations_members_data = await get_corporations_members_data(auth_token as string, lang, 'registered')
} catch (error) {
    data_fetching_error = error.message
}

const {
    corporations_members = [],
    main_count = 0,
    alliance_members_count = 0,
    unregistered_count = 0,
    corporations_unfiltered = {},
    corporations_members_count = {},
    total_members = 0,
    corporations_options = [],
} = corporations_members_data

const CORPORATIONS_MEMBERS_PARTIAL_URL = `${translatePath('/partials/corporations_members_component')}?status=registered`

import Viewport from '@layouts/Viewport.astro';

import PageDefault from '@components/page/PageDefault.astro';
import PageTitle from '@components/page/PageTitle.astro';

import FlexInline from '@components/compositions/FlexInline.astro';
import Flexblock from '@components/compositions/Flexblock.astro';

import CorporationMembersList from '@components/blocks/CorporationMembersList.astro';
import TextBox from '@components/layout/TextBox.astro';
import Input from '@components/blocks/Input.astro';
import ErrorRefetch from '@components/blocks/ErrorRefetch.astro';
import Button from '@components/blocks/Button.astro';
import NotificationBadge from '@components/blocks/NotificationBadge.astro';
import SelectFinder from '@components/blocks/SelectFinder.astro';

import MagnifierIcon from '@components/icons/buttons/MagnifierIcon.astro';

const page_title = t('corporations.members.page_title')
---

<Viewport
    title={page_title}
    components={{
        alert_dialog: true,
        confirm_dialog: true,
    }}
>
    <PageDefault
        cover={{
            image: "/images/alliance-cover.jpg",
            image_990: "/images/alliance-cover.jpg",
            animated: false,
            scrollable: true,
            overlay: true
        }}
        x-data={`{
            main_count: ${main_count},
            alliance_members_count: ${alliance_members_count},
            unregistered_count: ${unregistered_count},
            global_collapsed: [],
            corporations_members: ${JSON.stringify(corporations_members)},
            corporations_unfiltered: ${JSON.stringify(corporations_unfiltered)},
            corporations_filtered: ${JSON.stringify(corporations_unfiltered)},
            corporations_filtered_members_count: ${JSON.stringify(corporations_members_count)},
            total_members_filtered: ${total_members},
            total_members_unfiltered: ${total_members},
            corporations_options: ${JSON.stringify(corporations_options)},
            character_name: '',
            corporation_filter: '-1',
            filter() {
                if (this.character_name === '' && this.corporation_filter === '-1') {
                    this.corporations_filtered = this.corporations_unfiltered
                    this.corporations_filtered_memberss_count = ${JSON.stringify(corporations_members_count)}
                    this.total_members_filtered = this.total_members_unfiltered
                    return
                }

                this.corporations_filtered = {}
                this.corporations_filtered_memberss_count = {}
                this.total_members_filtered = 0

                this.corporations_members.forEach( (corporation) => {
                    if (this.corporation_filter !== '-1' && parseInt(this.corporation_filter) !== corporation.corporation_id) {
                        this.corporations_filtered[corporation.corporation_id] = []
                        this.corporations_filtered_members_count[corporation.corporation_id] = 0
                        return true
                    }

                    const members = corporation.members.filter( (member) => member.character_name.toLowerCase().includes(this.character_name.toLowerCase()) )
                    
                    if (members.length > 0) {
                        this.total_members_filtered += members.length
                        this.corporations_filtered[corporation.corporation_id] = members.map( (member) => member.character_id )
                        this.corporations_filtered_members_count[corporation.corporation_id] = members.length
                    }
                })
            },
        }`}
        x-effect="corporation_filter && filter()"
    >
        <FlexInline slot="header" justification='space-between' class="[ w-full ]">
            <PageTitle>
                {page_title}
            </PageTitle>
            <FlexInline>
                <Input type="text"
                    placeholder={t('filter')}
                    x-model="character_name"
                    x-on:keyup="filter()"
                >
                    <MagnifierIcon slot="icon" />
                </Input>
                
                <div class="[ hidden 2xl:block ]">
                    <SelectFinder
                        size={32}
                        placeholder={t('corporations')}
                        options="corporations_options"
                        model="corporation_filter"
                        width='280px'
                    />
                </div>
            </FlexInline>
        </FlexInline>

        <FlexInline slot="subheader">
            <Button
                color='transparent'
                href={translatePath('/alliance/corporations/members')}
                size='sm'
            >
                {t('members')}
                <NotificationBadge
                    x-text="alliance_members_count"
                    number={total_members ?? 0}
                />
            </Button>

            <Button
                href={translatePath('/alliance/corporations/members/unregistered')}
                size='sm'
            >
                {t('unregistered')}
                <NotificationBadge
                    color='fleet-yellow'
                    x-text="unregistered_count"
                    number={unregistered_count ?? 0}
                />
            </Button>
            
            <Button
                href={translatePath('/alliance/corporations/members/mains')}
                size='sm'
            >
                {t('mains')}
                <NotificationBadge
                    color='fleet-yellow'
                    x-text="main_count"
                    number={main_count ?? 0}
                />
            </Button>
        </FlexInline>

        <Flexblock gap="var(--space-xl)">
            <TextBox>
                <p>{t('corporations.members.leading_text')}</p>
            </TextBox>

            {data_fetching_error ?
                <ErrorRefetch
                    args={{
                        partial: CORPORATIONS_MEMBERS_PARTIAL_URL,
                        message: data_fetching_error,
                        delay: 0,
                    }}
                />
                :
                <CorporationMembersList corporations={corporations_members} />
            }
        </Flexblock>
    </PageDefault>
</Viewport>