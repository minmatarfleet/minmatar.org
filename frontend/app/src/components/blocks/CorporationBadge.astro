---
import { i18n } from '@helpers/i18n'
const { t } = i18n(Astro.url)

import type { CorporationBadgeProps } from '@dtypes/layout_components'
import { get_corporation_by_id } from '@helpers/api.eveonline/corporations'
import type { CorporationEvE } from '@dtypes/api.eveonline'
import { prod_error_messages } from '@helpers/env'

interface Props {
    corporation:    CorporationBadgeProps;
    border?:        boolean;
}

const {
    corporation,
    border = false,
} = Astro.props

let corporation_eve:CorporationEvE | null = null
let get_eve_corporation_error
if (!corporation?.name) {
    try {
        corporation_eve = await get_corporation_by_id(corporation.id)
    } catch (error) {
        get_eve_corporation_error = (prod_error_messages() ? t('get_eve_corporation_error') : error.message)
    }
}

const corporation_name = corporation?.name ?? (corporation_eve?.name ?? get_eve_corporation_error)
const size = corporation?.size ?? 'sm'
const logo_size = (size == 'sm' ? 32 : 64)
const description:string | false = corporation?.description ?? false

import { get_corporation_logo } from '@helpers/eve_image_server';
import FixedFluid from '@components/compositions/FixedFluid.astro'
import Flexblock from '@components/compositions/Flexblock.astro';

const gap = size === 'sm' ? 'var(--space-2xs)' : 'var(--space-s)'
---
<FixedFluid
    width={`${logo_size}px`}
    gap={gap}
    class="[ items-center ]"
>
    <picture class:list={{ border: border }}>
        <img loading="lazy" src={get_corporation_logo(corporation.id, logo_size)} width={logo_size} height={logo_size} />
    </picture>
    <Flexblock gap='0' class="[ w-full ]">
        {size == 'lg' ?
            <span>
                <span class="[ text-[var(--highlight)] max-w-fit ]">{corporation_name}</span>
            </span>
            <slot>
                <small>{description}</small>
            </slot>
            :
            <small class="[ max-w-fit ]">
                <a class="[ text-[var(--highlight)] ]">{corporation_name}</a>
            </small>
        }
    </Flexblock>
</FixedFluid>