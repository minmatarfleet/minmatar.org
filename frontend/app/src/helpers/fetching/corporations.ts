import { useTranslations } from '@i18n/utils';
import { get_user_character } from '@helpers/fetching/characters'
import type { EveCharacterProfile } from '@dtypes/api.minmatar.org'

const t = useTranslations('en');

import type { Corporation, CorporationApplication, CorporationType, CharacterCorp } from '@dtypes/api.minmatar.org'
import type { CorporationObject, CorporationStatusType, CorporationMembers, CharacterKind, CharacterBasic } from '@dtypes/layout_components'
import { get_all_corporations, get_corporation_by_id } from '@helpers/api.minmatar.org/corporations'
import { get_corporation_applications } from '@helpers/api.minmatar.org/applications'

export async function get_corporations_list_auth(access_token:string, user_id: number, corporation_type:CorporationType) {
    let api_corporations:Corporation[] = []
    let corporations:CorporationObject[] = []

    api_corporations = await get_all_corporations(corporation_type)

    corporations = await Promise.all(api_corporations.map(async (i) => add_status_to_corporation(access_token, i, user_id) ));
    
    return corporations
}

export async function get_corporations_list(corporation_type:CorporationType) {
    let api_corporations:Corporation[] = []
    let corporations:CorporationObject[] = []

    api_corporations = await get_all_corporations(corporation_type)

    corporations = api_corporations.map( (i):CorporationObject => {
        return {
            alliance_id: i.alliance_id,
            alliance_name: i.alliance_name,
            corporation_id: i.corporation_id,
            corporation_name: i.corporation_name,
            corporation_type: i.type,
            active: i.active,
            biography: i.biography,
            introduction: i.introduction,
            requirements: i.requirements,
            timezones: i.timezones,
            status: 'unauth'
        }
    } )
    
    return corporations
}

export async function get_corporation_list_by_id_auth(access_token:string, corporation_id:number, user_id: number) {
    let api_corporation:Corporation

    api_corporation = await get_corporation_by_id(access_token, corporation_id)

    return await add_status_to_corporation(access_token, api_corporation, user_id)
}

const add_status_to_corporation = async (access_token:string, api_corporation:Corporation, user_id:number) => {
    let corporation_applications:CorporationApplication[]

    const corporation:CorporationObject = {
        corporation_id: api_corporation.corporation_id,
        corporation_name: api_corporation.corporation_name,
        alliance_id: api_corporation.alliance_id,
        alliance_name: api_corporation.alliance_name,
        corporation_type: api_corporation.type,
        active: api_corporation.active,
        status: 'available',
        biography: api_corporation.biography,
        introduction: api_corporation.introduction,
        requirements: api_corporation.requirements,
        timezones: api_corporation.timezones,
    }

    try {
        corporation_applications = await get_corporation_applications(access_token, api_corporation.corporation_id)
    } catch (error) {
        corporation.status = 'error'
        return corporation
    }
    
    const user_application = user_id ? corporation_applications.find( (application) => application.user_id == user_id ) : undefined

    if (user_application !== undefined)
        corporation.status = user_application.status as CorporationStatusType

    return corporation
}

export async function get_user_corporation_id(user_id:number) {
    let user_character:EveCharacterProfile
    user_character = (user_id ? await get_user_character(user_id) : null)
    return !user_character ? null : (user_character?.corporation_id ?? null)
}

export async function get_all_corporations_members(access_token:string) {
    let api_corporations:Corporation[]
    let corporation_members:CorporationMembers[]

    api_corporations = await get_all_corporations('alliance')

    corporation_members = (await Promise.all(api_corporations.map(async (corporation) => 
        await get_all_corporation_members(access_token, corporation.corporation_id)
    )))

    return corporation_members
}

export async function get_all_corporation_members(access_token:string, corporation_id:number) {
    let api_corporation:Corporation
    let corporation_members:CorporationMembers

    api_corporation = await get_corporation_by_id(access_token, corporation_id)

    corporation_members = {
        corporation_id: api_corporation.corporation_id,
        corporation_name: api_corporation.corporation_name,
        active: api_corporation.active,
        type: api_corporation.type,
        members: []
    }

    corporation_members.members = (await Promise.all(api_corporation.members.map(async (api_member) => {
        const is_alt = (api_member.primary_character_id !== null)

        let member:CharacterKind = {
            character_id: api_member.character_id,
            character_name: api_member.character_name,
            registered: api_member.registered,
            exempt: api_member.exempt,
            is_main: !is_alt
        }
        
        if (is_alt) {
            member.main_character = {
                character_id: api_member.primary_character_id,
                character_name: api_member.primary_character_name,
            }
        }

        return member
    })))

    return corporation_members
}

export async function get_all_alliance_members(access_token:string) {
    let members:CharacterBasic[] = []    
    let api_corporations:Corporation[]
    let corporation_members:CorporationMembers[]

    api_corporations = await get_all_corporations('alliance')

    corporation_members = (await Promise.all(api_corporations.map(async (corporation) => 
        await get_all_corporation_members(access_token, corporation.corporation_id)
    )))

    corporation_members.map( (corporation) =>
        corporation.members.map( (member) => {
            members.push({
                character_id: member.character_id,
                character_name: member.character_name,
                corporation: {
                    id: corporation.corporation_id,
                    name: corporation.corporation_name,
                }
            })
        })
    )

    return members
}