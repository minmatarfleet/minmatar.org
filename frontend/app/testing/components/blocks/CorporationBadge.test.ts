import { experimental_AstroContainer as AstroContainer } from "astro/container";
import { expect, test, vi } from "vitest";
import CorporationBadge from "@components/blocks/CorporationBadge.astro";

import { get_corporation_by_id } from "@helpers/api.eveonline/corporations";
vi.mock("@helpers/api.eveonline/corporations");

test("CorporationBadge defaults", async () => {
  vi.mocked(get_corporation_by_id).mockResolvedValue({
    alliance_id: 43,
    ceo_id: 98,
    creator_id: 98,
    date_founded: new Date("2100-01-01"),
    description: "a great corp",
    faction_id: 1,
    home_station_id: 65,
    member_count: 91,
    name: "chicken attack",
    shares: 12,
    tax_rate: 1,
    ticker: "CHICK",
    url: "chicken.com",
    war_eligible: true,
  });
  const container = await AstroContainer.create();
  const result = await container.renderToString(CorporationBadge, {
    props: {
      corporation: {
        id: 12,
        name: "chicken attack",
      },
    },
  });

  expect(result).toMatchSnapshot();
});
