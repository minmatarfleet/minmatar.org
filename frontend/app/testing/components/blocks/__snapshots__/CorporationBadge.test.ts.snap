// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CorporationBadge defaults 1`] = `"<div class="fixed-fluid [ items-center ]" data-mod data-astro-cid-34go4nvk style="--fixed_fluid_gap: var(--space-2xs);--fixed_fluid_width: 32px;--fixed_fluid_breakpoint: 50%;">  <picture> <img loading="lazy" src="https://images.evetech.net/corporations/12/logo?size=32" width="32" height="32"> </picture> <div class="flex-block [ w-full ] gap-0" data-mod style="--flex_block_gap: 0;"> <small class="[ max-w-fit ]"> <a class="[ text-[var(--highlight)] ]">chicken attack</a> </small> </div>   </div>  "`;
